Add Pipe Commands extension

pomdtr opened the pull request:
## Description

Pipe commands allows you to send your selection to a script command, and act on it. Example usage:
- send selected file by mail
- transform user selection to lowercase
- search selected word in google...

Some scripts are provided as examples in the assets folder.

## Type of change

<!-- Please choose one option and delete others that are not relevant. -->

- New extension

## Screencast

![Screen Recording 2022-01-25 at 00 26 06](https://user-images.githubusercontent.com/17577332/151875468-8d9b7ca5-4c49-4f10-8e23-1bb9d3ff5900.gif)

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

fturcheti commented:
Hey, @pomdtr! I was testing some icons with more complex forms, trying to indicate something being transformed or piped, but I didn't like the results. Icons in Raycast are too small, so simpler forms tend to fit better. 

This is the one I liked the most: 

![icon-pipe-to-command](https://user-images.githubusercontent.com/365403/151036297-ab39de1b-8510-4708-9af7-e675d7be870d.png)

What do you think?

fturcheti commented:
From my understanding of the way this extensions is designed, it is not possible to create a command (a pipe command) that presents a list of actions to the user, right? 

The use case I'm thinking of is the one of my extension "Open With App". It works like this:

1. You select a file in Finder (or multiple files)
2. Launch Raycast
3. Select the command "Open With…"
4. Select an application from a list of applications to open the file with

pomdtr commented:
I love the icon, thanks a lot !
I have decided to hardcode the openWithApp action :)

fturcheti commented:
> I actually want to create another extension allowing to generate list from script commands, using a syntax inspired by alfred script filters.

Nice!

pomdtr commented:
Not sure why the CI returns an error here, `npm run lint` does work on my side.

fturcheti commented:
Hey, @pomdtr. When using pipe commands for texts transformations, I'd prefer the result to be copied to the clipboard instead of direct pasted. 

I know this can be changed by editing the commands source code, but I'd like to know your thoughts on which should be the suggested default for this transformations. I think that putting the transformed text in the clipboard is more flexible—you can `cmd+v` it to substitute the original text or use it whatever you want, because you have it "saved" on your clipboard.

pomdtr commented:
> Hey, @pomdtr. When using pipe commands for texts transformations, I'd prefer the result to be copied to the clipboard instead of direct pasted. 
> 
> I know this can be changed by editing the commands source code, but I'd like to know your thoughts on which should be the suggested default for this transformations. I think that putting the transformed text in the clipboard is more flexible—you can `cmd+v` it to substitute the original text or use it whatever you want, because you have it "saved" on your clipboard.

I agree, I think that all text commands should allow to either replace the selection or go to the clipboard (with a pref to configure the default action).

We should never have to use pbcopy in the script code.

pomdtr commented:
> @pomdtr I'll list here other suggestions/thoughts:
> 
> * It would be nice to have a "Edit Pipe Command" command that would list your commands with a default action to "Edit" them with your default text editor.
> * Besides acting on the selection, it could fallback to act on the clipboard.
> * Is it possible to edit or remove built-in pipe commands? This options are only shown for the user created ones.
> * If I start to rename a file in Finder and invoke the command with the filename selected, the extension detects the selection as "file" not as "text". This is not the expected behavior.

I agree, I think the extension will just copy the default scripts to the supportFolder on the first run (all pipe commands will be editable / deletable).

pomdtr commented:
> If I start to rename a file in Finder and invoke the command with the filename selected, the extension detects the selection as "file" not as "text". This is not the expected behavior.

Unfortunately I did not find a way to correct this with raycast current API. The current behavior is the best I was able to achieve.

fturcheti commented:
> Unfortunately I did not find a way to correct this with raycast current API. The current behavior is the best I was able to achieve.

I understand. But this got me thinking: maybe it'd be a good idea to have a "Pipe Clipboard to Command" command. I'm not sure if I'd use more "pipe from clipboard" or "pipe from selection", but "pipe from clipboard" would work for a situation like this file renaming—I could copy the name, invoke "Pipe Clipboard to Command", apply a pipe command to transform the text, and get the result in my clipboard to paste over the original name.

Would a "Pipe Clipboard to Command" work for files too?

pomdtr commented:
The extension is currently defaulting to the clipboard if no item is selected. But maybe I should introduce a real `pipe clipboard` command ?

fturcheti commented:
> The extension is currently defaulting to the clipboard if no item is selected. But maybe I should introduce a real pipe clipboard command ?

Maybe instead of using the clipboard as a fallback in the "Pipe Selection to Command", it's a better idea to have a "Pipe Clipboard to Command"—it's more explicit on what's going on. Also, there are situations like this:

1. You select some text
2. Invoke the "Pipe Selection to Command"
3. The selection detection fails (on an unsupported app, or for whatever reason)
4. The extension use the clipboard as a fallback
5. The extension pastes the transformed text from the clipboard over the original selection, although the user would like to transform the selection content, not the clipboard one

I know I have suggested the clipboard as a fallback, but thinking about it a bit more, it doesn't seem to be a good idea. Another command, to explicitly act on the clipboard seems to be a better approach.

What do you think? Does it make sense to you?

_
P.S. I just discovered that Raycast can't detect Neovim selections on the terminal, and that's my editor of choice. So I'd use the "Pipe Clipboard to Command" more than I initially thought.

pomdtr commented:
Ok this should be ready for review !
@fturcheti I would love to get your feedback on the last version.

fturcheti commented:
Hey, @pomdtr! I'll keep testing the extension, but here are some things that already popped up: 

- I've got an "Invalid Pipe Command" for a ".DS_Store" file (screenshot below). I think the fix for it is just a matter of filtering them out of this listing, right?
- "Pipe Clipboard to Command" is not working with files. Is this a limitation from the API?
- I miss an "Edit Pipe Command" that would list all your pipe commands (maybe in with list sections for each type—text, file and url). It would make it easy to manage them. 
- It'd be hand to have a shortcut assigned  to the "Pipe Output" action (maybe `⌘ + ]`)

![Screen Shot 2022-01-31 at 23 30 24](https://user-images.githubusercontent.com/365403/151976756-d46a2c6d-f629-47d1-9e27-4071d60ca54a.png)

pomdtr commented:
Hey @fturcheti thanks for the feedback !

1. Yeah I think all dotfiles should be ignored
2. I have no way to find out that the current clipboard is a file using the pbpaste command :/
3. I agree. I think I will remove the invalid pipe command from the default menu, and display them here
4. I agree, I think I will use `cmd+shift+\` (close to a shell pipe `|`)

pernielsentikaer commented:
What also could be an awesome idea, is that the first launch showed a `detail view` with information on how to use it somehow

fturcheti commented:
@pomdtr Another thing that I noticed makes me less confident with the UX: it doesn't make it clear what's the thing your piping into a command. To tackle this with the current UI components from the API, I see two options:

1. Use a List section title
2. Or a Toast with the style ToastStyle.Animated

This component would display the text (or part of the text…) you have selected, or the URL, or the filename, depending on the thing your piping. If there's nothing selected, it could display a notice while still making the list available for managing pipe commands — an error message would appear only if the user tried to run an action on an empty selection.

What do you think?

fturcheti commented:
@pomdtr And just a small UI tweak: you could use the "globe" icon for URL commands, to differentiate them from text commands.

pomdtr commented:
Interesting idea. I removed the url type I felt like it did not bring enough values compared to the text type

pomdtr commented:
I feel like it will be easier to show the user the selection once the dual panel will be available in the raycast API

pomdtr commented:
> "Pipe Clipboard to Command" is not working with files. Is this a limitation from the API?

I found a way to achieve this using applescript :)

fturcheti commented on extensions/pipe-commands/src/pipe-to-command.tsx:
```diff
@@ -43,7 +43,7 @@ export function PipeCommands(props: { input: PipeInput }): JSX.Element {
         <List.Item
           icon={Icon.Document}
-          title="Open With"
+          title="Open with..."
```
I think it's better to name it "Open With…", as this is the way Raycast name this action in the built-in File Search extension.

fturcheti commented on extensions/pipe-commands/src/manage-pipe-commands.tsx:
```diff
@@ -0,0 +1,50 @@
+
+  return (
+    <List isLoading={typeof state == "undefined"}>
+      <List.Section title="text">
```
I think "Text Commands" would be a better name for the section

fturcheti commented on extensions/pipe-commands/src/manage-pipe-commands.tsx:
```diff
@@ -0,0 +1,50 @@
+          <PipeCommand key={cmd.path} command={cmd} reload={loadCommands} />
+        ))}
+      </List.Section>
+      <List.Section title="file">
```
The same way, "File Commands" would fit better here

fturcheti commented on extensions/pipe-commands/src/manage-pipe-commands.tsx:
```diff
@@ -0,0 +1,50 @@
+  const fileCommands = state?.commands.filter((command) => command.metadatas.input.type === "file");
+
+  return (
+    <List isLoading={typeof state == "undefined"}>
```
In regards to the order of the sections, I'd use this order: "invalid", "file", "text". I'd display "invalid" at the top so the user could notice it and act to solve the issue with these commands.  And I'd put the "file" section before the "text" one cause I think these should be alphabetic ordered.

pomdtr commented:
Thanks for the feedback @fturcheti 
I adressed all of these :)

fturcheti commented:
Hey @pomdtr! What do you think about custom icons for pipe commands? Take a look:

<img width="750" alt="Screen Shot 2022-02-16 at 09 41 08" src="https://user-images.githubusercontent.com/365403/154266600-a7c532e0-76e4-4047-892b-f7333d4e60f1.png">

<img width="605" alt="Screen Shot 2022-02-16 at 09 41 21" src="https://user-images.githubusercontent.com/365403/154266622-fe9dc4ef-faa2-4aac-a196-4dfd9ded62ce.png">

fturcheti commented:
> I added this feature in https://github.com/raycast/extensions/commit/db57c0b95776f8fbf6fe5146da137f0698fa9daf :)

Nice @pomdtr ! I saw your comment suggesting a PR but I was not able to do it properly at that time. Also, I'm a completely newbie with Typescript. I'm starting to "get it" looking though these Raycast extensions, so it's nice to see how you solved this.

In regards to the icons itself, the good thing about using the built-in ones is that they keep things clean, concise, and they work with both light and dark themes. When providing external icons, you can end up with contrast problems (the Github favicon is a problematic one, as it will disapear in the dark theme). Script-commands, for instance, have an "icon" and an "iconDark" property to tackle this issue.

pomdtr commented:
> > I added this feature in https://github.com/raycast/extensions/commit/db57c0b95776f8fbf6fe5146da137f0698fa9daf :)
> 
> Nice @pomdtr ! I saw your comment suggesting a PR but I was not able to do it properly at that time. Also, I'm a completely newbie with Typescript. I'm starting to "get it" looking though these Raycast extensions, so it's nice to see how you solved this.
> 
> In regards to the icons itself, the good thing about using the built-in ones is that they keep things clean, concise, and they work with both light and dark themes. When providing external icons, you can end up with contrast problems (the Github favicon is a problematic one, as it will disapear in the dark theme). Script-commands, for instance, have an "icon" and an "iconDark" property to tackle this issue.

You are right, I will add support for the iconDark property.

fturcheti commented:
And just another comment in regard to this: in my dirty hack I kept the different default icons for each type of command. So if a `text` command didn't have a custom icon, it would default to `Icon.Text`, and a `file` command would default to `Icon.Document`. This seems to be a good thing to keep from the implementation prior to custom icons. With a generic default we loose this basic differentiation between the types of commands. What do think?

pomdtr commented:
> And just another comment in regard to this: in my dirty hack I kept the different default icons for each type of command. So if a text command didn't have a custom icon, it would default to Icon.Text, and a file command would default to Icon.Document. This seems to be a good thing to keep from the implementation prior to custom icons. With a generic default we loose this basic differentiation between the types of commands. What do think?

Great suggestion ! Fixed in b479fc2

thomaspaulmann commented:
What's the latest on this? Looks like folks can't wait for it ;-) 

https://twitter.com/srikat/status/1496569084154384385

pomdtr commented:
The extension is waiting for review. I stil do some minor improvements from time to time as I'm using it daily.

mathieudutour reviewed (commented):
It looks amazing, really good job!

I just have a comment about supporting `@raycast.argument1` instead of `@raycast.input` so that some Script Commands can work out of the box with this extension.

Let me know if you want to do that, otherwise I'll merge it

mathieudutour commented:
Another idea I had while playing with it: it'd be nice to have "pipelines", some sort of scripts than call some others in row

pomdtr commented:
> Another idea I had while playing with it: it'd be nice to have "pipelines", some sort of scripts than call some others in row

I think this usecase should be handled directly by the shell by calling multiple other pipe commands from the pipe command.

pomdtr commented:
> It looks amazing, really good job!

> I just have a comment about supporting @raycast.argument1 instead of @raycast.input so that some Script Commands can work out of the box with this extension.

> Let me know if you want to do that, otherwise I'll merge it

I added support for the `raycast.argument1` field, and dropped `raycast.input`. For convenience, the input is passed both as stdin and as the script first argument.

mathieudutour reviewed (approved):
Neat, let's go 🚀
